File: Linked_List/Python/Reorder List.py
# Problem link: https://leetcode.com/problems/reorder-list/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        prev, curr, nxt = None, slow.next, None

        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        
        fast = head

        while prev:
            nxt = prev.next
            fast.next, prev.next = prev, fast.next
            fast = prev.next
            prev = nxt
        
        fast.next = prev

File: Linked_List/Java/Reorder List.java
// Problem link: https://leetcode.com/problems/reorder-list/description/
// Time: O(n)
// Space: O(1)
class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head, fast = head.next;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prev = null, curr = slow.next, nxt = null;

        while(curr != null) {
            nxt = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nxt;
        }

        fast = head;

        while(prev != null) {
            nxt = prev.next;
            prev.next = fast.next;
            fast.next = prev;
            fast = prev.next;
            prev = nxt;
        }

        fast.next = prev;
    }
}
